Ruby SDK - Phase 1 - Initial Comment Period

cretz opened the pull request:
# Summary

* This is the proposal for the first phase of the Ruby SDK.
* See it [rendered](https://github.com/cretz/temporal-proposals/blob/ruby-phase-1/ruby/sdk-phase-1.md).
* Much of this is inspired by the current [Temporal Ruby SDK](https://github.com/temporalio/sdk-ruby) (and its [proposals](https://github.com/temporalio/proposals/tree/master/ruby/previous) which were moved to a sub directory) and our other SDKs.
* Use this PR for comments/discussion. Don't be afraid of making too many comments; if it gets too noisy, I'll just create another PR.
* Any points of contention will be discussed in-person by the SDK team and a decision will be made.

bergundy commented on ruby/sdk-phase-1.md:
```diff
@@ -0,0 +1,439 @@
+
+These are the general requirements of the project
+
+* Ruby >= 3.1
```
Wondering what we've done to validate this is acceptable.

josh-berry replied:
All older versions of Ruby are EOL; I don't think we need to support EOL software: https://www.ruby-lang.org/en/downloads/

cretz replied:
Correct: https://en.wikipedia.org/wiki/History_of_Ruby. But all of this is still pending user feedback and so we'll likely be surveying the community on how acceptable they consider this.

bergundy commented on ruby/sdk-phase-1.md:
```suggestion
* No compatibility with _any_ existing Ruby SDK
```

cretz replied:
Not sure the emphasis is needed

bergundy commented on ruby/sdk-phase-1.md:
I've never seen the need to create more than a single runtime in our SDKs. Seems like we can avoid this extra API surface to expose accepting a runtime and just have all constructs reference a singleton.

cretz replied:
They will reference a lazy global which is basically a singleton. This is deemed acceptable in other core-based SDKs and so should be here too, no need to make Ruby special.

josh-berry replied:
Yeah I'm fine with it being a singleton/lazy global under the hood. I just don't want us to design ourselves into a corner.

bergundy commented on ruby/sdk-phase-1.md:
namespace defaults to `default`?

cretz replied:
Yes

cretz replied:
This can definitely be argued, but I think enough of our SDKs default to "default" and some prod clusters do use "default" that yeah it's not a big deal.

cretz replied:
We are going to update this to require the namespace. The default of `default` is unreasonable for cloud and many others.

bergundy commented on ruby/sdk-phase-1.md:
Don't know what the Ruby standard is here but as you know in our other SDKs we've chosen the longer name.

cretz replied:
The Ruby standard here is to keep it shorter if already in the module name. Ruby is the only language I can think of that does this commonly. While you can require a file, there is no "importing" a module, it must always be qualified (but people often alias it).

bergundy commented on ruby/sdk-phase-1.md:
In other SDKs, `heartbeat`, `log`, `info`, etc... is on the `activity` module.
Why not do that here too? E.g:

```
Temporalio::Activity::Context.current

Temporalio::Activity::heartbeat
Temporalio::Activity::info
Temporalio::Activity::log
```

cretz replied:
Because `Temporalio::Activity` is the class that is being extended by the activity implementation and we don't want people to think that this is only usable in that extension nor do we want to clutter the inherited class with methods like these when they are part of the context.

antlai-temporal commented on ruby/sdk-phase-1.md:
What about JSON Protobuf, I'm assuming not supported...

Also, what are the implications for cross-language integration. For example, an activity in python that returns a json serialized object called from a Ruby workflow. Is the workaround to manually create the ruby object from json?

cretz replied:
> What about JSON Protobuf, I'm assuming not supported...
JSON protobuf will be supported out of the box

> Also, what are the implications for cross-language integration. For example, an activity in python that returns a json serialized object called from a Ruby workflow. Is the workaround to manually create the ruby object from json?

Yes, that will be a Ruby hash (just like it would be a JS object). If you want to convert it to a Ruby class you would have to do that yourself. We can consider more advanced JSON-to-class conversion mechanisms later, but by default Ruby doesn't really have any without the "addition" approach which is not very cross-language compatible.

antlai-temporal commented on ruby/sdk-phase-1.md:
Is there a way to abstract the low level implementation (fiber vs thread pool) from how the cancellation is reported to the activity code. Ideally, we would like to change the implementation without modifying application code...

Sushisource replied:
+1 - would be nice to just have `Temporalio::Activity::Context.raise`

cretz replied:
> Is there a way to abstract the low level implementation (fiber vs thread pool) from how the cancellation is reported to the activity code

It is abstracted from a user POV. This discusses what _our_ SDK code will do when these happen. From a user POV it's just an exception that occurs.

Sushisource replied:
Ah, cool that makes sense, that wasn't immediately clear to me.

antlai-temporal commented on ruby/sdk-phase-1.md:
The temptation for programmers is to add state (in the activity object) that lingers across multiple activity invocations, something really fragile... Is there a way to freeze that activity object so that it can only be used in a pure functional style?

Sushisource replied:
Yeah, that's my concern too. I think instantiating them per-invocation is probably safer, and sharing things like dependencies could be done w/ either static fields or simply by passing them in each time.

cretz replied:
Technically in .NET if you use DI w/ scoped activities it does instantiate them per activity task, but overall in our SDKs you are allowed to share object/outside-of-function state across activity invocations without using globals. I am a hesitant to make Ruby the special case of forcing globals to share state.

But we can probably have ways to let the user choose. For instance, we can accept two ways of providing an activity: providing an instance and providing a class. I wouldn't want the activity class form to instantiate per attempt _by default_ because it'd be confusing to have whether you did instance vs class to have different lifetimes. But we could support some setting on the activity class to say they want it per attempt and hope they have no constructor parameters. Or we'd have to devise some DI because in general it is difficult for Temporal to control lifetimes of user activity objects due to statefulness needs.

Sushisource commented on ruby/sdk-phase-1.md:
I think classes makes good sense for Ruby. The typing is good and Ruby is all about the OOP nonsense. The biggest downside is users thinking that storing class state is somehow magically going to be synchronized across workers or something, but I'm not hugely concerned about that.

josh-berry commented on ruby/sdk-phase-1.md:
This wouldn't be for an MVP, but could we eventually provide both if we wanted to? It doesn't seem like there's an obvious trade-off winner here.

Another plus in favor of functions is they are lighter-weight to define; you don't have to make a class just to wrap what is effectively a function anyway.

cretz replied:
We can technically do this, and could technically provide activities as classes/interfaces/structs in every other SDK but we have tried to stick with a single way that is best with the language. I think we should only do one here.

josh-berry replied:
Yeah if we want to pick a single way, then I'm in favor of doing the most-flexible thing. Long as we have room to add the other way later if we learn that users are really opinionated about it.

Sushisource commented on ruby/sdk-phase-1.md:
I think that's best if doable

cretz replied:
It is doable, it just means you have to instantiate the worker in the async context as well

Sushisource reviewed (approved):
:gem:

josh-berry commented on ruby/sdk-phase-1.md:
So activities are essentially singletons? Bleh, that seems harder to reason about and rather counter-intuitive to anyone new to Temporal. Is this a common pattern in Ruby?

cretz replied:
Up to a user, they own the lifetime of these objects not us. They could instantiate different ones for different workers or share across workers. But yes, they live for the life of the worker which makes them long-lived usually (same with something like a database client). There is not a single common patterns for this kind of thing in Ruby that I am aware of. The current Temporal Ruby SDK does instantiate per activity task and didn't take into account the need to share things like a database client which is common.

josh-berry commented on ruby/sdk-phase-1.md:
Are we also thinking about workers that have a combination of sync and async activities?

cretz replied:
Yes, this may be common (and is common in Python). This is just saying _if_ any are async we make sure we're in an async environment. Sync activities will always work.

cretz commented on ruby/sdk-phase-1.md:
But in the case of `SearchAttributes` (plural) and `Metric` (singular), they may in fact be classes (the former being the search attribute collection and the latter being the base metric class).

Sushisource replied:
Yeah, I think you maybe misread me, I said because they *are* classes

cretz replied:
I did, sorry!

cretz commented on ruby/sdk-phase-1.md:
> Does either sync or async depend on the other? Or are the implementations independent?

Not following what is meant by "depend on the other", but they are independent for the most part. Basically it's just the difference of starting a new thread or a new fiber when invoking the activity, not much else changes.

josh-berry replied:
> Not following what is meant by "depend on the other",

Is sync implemented in terms of—or does it reuse code written for—async, or vice versa? I think you've answered my first question here though.

For my second question, I'm still trying to understand what is MVP-shaped and what isn't—e.g. would we be comfortable doing a first/early release with one but not the other?

cretz replied:
Sorry, missed this question. This is integral behavior and the same code will work async/sync so there is no real separating it.

cretz commented on ruby/sdk-phase-1.md:
Sure, though we should do the same for https://java.temporal.io/ too (nothing there, should redirect to https://www.javadoc.io/doc/io.temporal) and https://go.temporal.io/ too (redirects to temporal.io, should redirect to https://pkg.go.dev/go.temporal.io/sdk maybe, though there is https://pkg.go.dev/go.temporal.io/api and https://pkg.go.dev/go.temporal.io/server too). Want to start the effort there?

josh-berry replied:
Sure, will make a note for later (I don't think this is urgent)

dandavison commented on ruby/sdk-phase-1.md:
```diff
@@ -0,0 +1,439 @@
+    else if we can help it.
+  * 💭 Why? Our dependencies as a library have transitive effects on our users, so we have a responsibility to be
+    lightweight.
+* Must be both thread-blocking/thread-safe and async-capable
```
What do you mean exactly by "async-capable"? (The term is used in a few places)

cretz replied:
In Ruby there are threads and fibers (basically coroutines). By async-capable I mean fiber-capable. So you can use https://github.com/socketry/async and, say, use the client to wait on 100 workflow results without using a thread each. But many Ruby users are totally fine using a thread each and we need to support that too.

This is very much like Python, we just chose not to provide a synchronous version of the Python client (though we still could if we wanted).

cretz commented on ruby/sdk-phase-1.md:
In most of our SDK languages there are better ways to run multiple workers at once without eating a thread per worker. JS is always single threaded, Python has `asyncio.gather`, Go has goroutines, and .NET has tasks. In Java we have a `WorkerFactory` that lets you run multiple workers at once because it lacks good language-level reuse-thread capabilities. Ruby has the same problem. So I was thinking a `WorkerPool` or similar that can run many workers in a single blocking run call.

drewhoskins-temporal commented on ruby/sdk-phase-1.md:
Context related to this topic, ruby @ Stripe was single-threaded by default.  Multithreaded carve-outs did happen but were relatively unusual.
This meant some libraries weren't designed to be thread-safe and used things like class variables.
Default was therefore to have many workflow pollers but only one activity poller per process, which was inefficient, but we added a config flag folks could turn on to make their workers run multiple activity pollers.  I really wanted to make the multi-poller version the default but didn't do so before I left.

cretz replied:
This is why we plan to also support single-threaded/coroutine-based implementations. It's important to us at Temporal that people be able to do a lot of concurrent things with the SDK in just a single thread. Granted the current design is queue-pop-based which makes it work with fiber-based async. I wasn't planning on supporting other async approaches (e.g. event machine) but if there is enough desire for pluggable-non-fiber-based async, we can look into that.

> Multithreaded carve-outs did happen but were relatively unusual.

These carve-outs were likely required for to use the existing Ruby SDKs.

drewhoskins-temporal replied:
On the carve-outs, I'm referring to the "pre-temporal world."  Later, with the old SDK, we simply limited to one activity poller, which fixed any race conditions.  It was a cautious approach, as most customers probably would have been fine, but there was FUD.

cretz commented on ruby/sdk-phase-1.md:
Fixed

cretz commented on ruby/sdk-phase-1.md:
Fixed

cretz commented on ruby/sdk-phase-1.md:
Fixed

cretz commented:
I have updated the proposal with the following:

* Clarify why we need typing at all
* Use `Temporalio::SearchAttributes` as both the collection and class under which all nested search attribute things reside
* Use `Temporalio::Metric` as both the base metric and class under which all nested metric things reside
* Clarified use of `Options` structs on client/connection
* Clarified there will be a `CloudOperationsClient`
* Made `namespace` a required field instead of defaulting to `default`
* Changed `Temporalio::Converter` module to `Temporalio::Converters`
* Changed `Temporalio::Converters::PayloadConverter` to be a simple base class instead of having it impl composite, and created `Temporalio::Converters::PayloadConverter::Composite`
* Updated activity lifecycle to say we will instantiate activities on each attempt if registered as a class, or we will just call method on existing instance if registered as an instance
* Clarified that activity executors can validate activities at registration time instead of runtime, which means that workers have to be _created_ in an async context if they have any async-context-needed activities (as opposed to just run in an async context)

cretz commented:
Closing this PR in favor of a new one for a fresh round of comments. Will link back to this one for posterity, but unaware of any unresolved issues here. Anyone here can feel free to comment on the new one.
